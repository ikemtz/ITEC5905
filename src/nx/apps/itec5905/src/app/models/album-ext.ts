
import { IAlbum } from "../../models/artists-webapi";
export interface IAlbumExt extends IAlbum {
  image?: string | null;
}