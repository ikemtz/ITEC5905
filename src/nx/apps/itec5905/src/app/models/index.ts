export * from './album-ext';
export * from './album-upsert';
export * from './artist-ext';
export * from './artist-upsert';
export * from './song-ext';
export * from './song-upsert';
