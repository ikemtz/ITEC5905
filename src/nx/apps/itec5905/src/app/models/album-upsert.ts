import { IMediaUploadRequest } from "../../models/media-webapi";
import { IAlbumUpsertRequest, } from "../../models/artists-webapi";

export interface IAlbumUpsert {
  album: IAlbumUpsertRequest,
  picture?: IMediaUploadRequest,
}