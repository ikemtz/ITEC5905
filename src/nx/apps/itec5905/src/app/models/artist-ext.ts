
import { IArtist } from "../../models/artists-webapi";
export interface IArtistExt extends IArtist {
  image?: string | null;
}