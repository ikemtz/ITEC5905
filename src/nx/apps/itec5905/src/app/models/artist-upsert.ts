import { IMediaUploadRequest } from "../../models/media-webapi";
import { IArtistUpsertRequest } from "../../models/artists-webapi";

export interface IArtistUpsert {
  artist: IArtistUpsertRequest,
  picture?: IMediaUploadRequest,
}