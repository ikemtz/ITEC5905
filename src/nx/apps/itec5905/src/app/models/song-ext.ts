
import { ISong } from "../../models/artists-webapi";
export interface ISongExt extends ISong {
  image?: string | null;
}