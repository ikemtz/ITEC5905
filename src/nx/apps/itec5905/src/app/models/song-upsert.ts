import { ISongUpsertRequest } from "../../models/artists-webapi";
import { IMediaUploadRequest } from "../../models/media-webapi";

export interface ISongUpsert {
  song: ISongUpsertRequest,
  picture?: IMediaUploadRequest,
  audio?: IMediaUploadRequest,
}