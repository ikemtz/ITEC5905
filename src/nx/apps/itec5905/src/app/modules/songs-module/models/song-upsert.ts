import { ISongUpsertRequest } from "apps/itec5905/src/models/artists-webapi";
import { IMediaUploadRequest } from "apps/itec5905/src/models/media-webapi";

export interface ISongUpsert {
  song: ISongUpsertRequest,
  picture?: IMediaUploadRequest,
  audio?: IMediaUploadRequest,
}