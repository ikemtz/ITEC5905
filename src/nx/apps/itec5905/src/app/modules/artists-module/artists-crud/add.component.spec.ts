import { CUSTOM_ELEMENTS_SCHEMA, NO_ERRORS_SCHEMA } from '@angular/core';
import { waitForAsync, ComponentFixture, TestBed } from '@angular/core/testing';
import { ReactiveFormsModule } from '@angular/forms';
import { NoopAnimationsModule } from '@angular/platform-browser/animations';
import { DropDownsModule } from '@progress/kendo-angular-dropdowns';
import { createDataEntryMockFacade } from 'imng-kendo-data-entry/testing';
import { mockConsoleError, mockConsoleGroup, mockConsoleWarn, readFirst } from 'imng-ngrx-utils/testing';
import { of } from 'rxjs';
import { ArtistGenreUpsertRequestFormGroupFac, IArtist, createTestArtist, createTestArtistGenre } from '../../../../models/artists-webapi';

import { ArtistAddComponent } from './add.component';
import { ArtistCrudFacade } from './crud.facade';


export function createMockArtistFacade() {
  return {
    currentEntity$: of({}),
    pictures$: of([
      { id: 'abc', blob: 'abc', type: 'abc', },
      { id: 'xyz', blob: 'xyz', type: 'xyz', },]),
    loadPictures: jest.fn(),
  };
}

describe('ArtistAddComponent', () => {
  let component: ArtistAddComponent;
  let fixture: ComponentFixture<ArtistAddComponent>;
  let facade: ArtistCrudFacade;
  let consoleWarnMock: jest.SpyInstance<void>;
  let consoleGroupMock: jest.SpyInstance<void>;

  beforeEach(waitForAsync(() => {
    consoleWarnMock = mockConsoleWarn();
    consoleGroupMock = mockConsoleGroup();
    TestBed.configureTestingModule({
      declarations: [ArtistAddComponent],
      imports: [ReactiveFormsModule, NoopAnimationsModule, DropDownsModule,],
      providers: [{ provide: ArtistCrudFacade, useValue: createDataEntryMockFacade(createMockArtistFacade()) }],
      schemas: [CUSTOM_ELEMENTS_SCHEMA, NO_ERRORS_SCHEMA],
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(ArtistAddComponent);
    component = fixture.componentInstance;
    facade = TestBed.inject(ArtistCrudFacade);
    fixture.detectChanges();
  });

  afterAll(() => {
    component.ngOnDestroy();
    consoleWarnMock.mockRestore();
    consoleGroupMock.mockRestore();
  });

  test('should create', () => {
    expect(component).toBeTruthy();
  });

  test('should save', () => {
    component.initForm();
    component.addEditForm?.patchValue(createTestArtist());
    const genreForm = ArtistGenreUpsertRequestFormGroupFac();
    genreForm.patchValue(createTestArtistGenre());
    component.addEditForm.controls.genres?.controls.push(genreForm);

    let item: IArtist | undefined;
    facade.saveNewEntity = jest.fn(x => (item = x));
    facade.updateExistingEntity = jest.fn();
    expect(component.getFormErrors()).toStrictEqual([]);
    component.onSubmit();
    expect(facade.saveNewEntity).toBeCalledTimes(1);
    expect(facade.updateExistingEntity).toBeCalledTimes(0);

    expect(item).toMatchSnapshot();
  });

  /**
   * Note: if this test fails, then you're missing validators in your forms.
   * Using form validators is typically a good idea.
   */
  test('should not save', () => {
    const consoleErrorMock = mockConsoleError();
    component.addEditForm?.patchValue({});
    component.onSubmit();
    expect(facade.saveNewEntity).toBeCalledTimes(0);
    expect(facade.updateExistingEntity).toBeCalledTimes(0);
    consoleErrorMock.mockRestore();
  });

  test('should cancel', () => {
    facade.clearCurrentEntity = jest.fn();
    component.cancel();
    expect(facade.clearCurrentEntity).toBeCalledTimes(1);
  });

  test('should support Picture filters', async () => {
    component.handlePictureFilter('xy');
    const result = await readFirst(component.pictures$);
    expect(result).toMatchSnapshot();
  });
});
