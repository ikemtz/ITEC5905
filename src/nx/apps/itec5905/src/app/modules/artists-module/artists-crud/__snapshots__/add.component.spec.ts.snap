// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArtistAddComponent should save 1`] = `
{
  "email": "artist@email.org",
  "genres": null,
  "id": undefined,
  "name": "NAME",
  "stageName": "STAGE_NAME",
}
`;

exports[`ArtistAddComponent should support Picture filters 1`] = `
[
  {
    "blob": "xyz",
    "id": "xyz",
    "type": "xyz",
  },
]
`;
